### DISCUSSION TITLE
Markups Revisited

### DISCUSSION THREAD

**Discussion by kirillta** [happy-travel/docs#17] [Ideas]
The current implementation of markups appears to be complicated for end-users, so we have to redesign its management and behavior. Based of a user's feedback, I suggest the following:

1. _Merge global and agency-level markups._ Global markups will become agency-level markups to get an ability to apply discounts in a new manner. Where will be only agency, sub-agency, and agent levels.
2. _Each level has a base (starting) value._ A final level value can't lower the base value.
3. _One can only add or subtract a markup._ There will be an addition of percent instead of a multiplication:
```
x × 1.02 = 1.02x // old way
x + 2% = 1.02x // new way
```
4. _A negative markup is a discount._ Discount applies to the base value as well as markups: `x + 3% – 1% = x + 2%`.
5. _Percent applies to percent, fixed to fixed._ And fixed ones apply at the very end, and because of its nature the fixes implies such markup should not affect percentage, like a price for a transfer for instance:
```
(x + 2% + 15USD) + 3% + 3USD = 
= 1.02x + 3% + 18USD = 
= 1.02x × 1.03 + 18USD = 
= 1.0506x + 18USD
```
6. _A markup ordering becomes not necessary at this point._
7. _A markup still may be applied on agency/sub-agency/agent in one direction and on region/country/locality/accommodation on another._
8. _Only an admin with a certain permission able to manage markups at the admin panel._
9. _All manipulations must be logged to an audit table._

**Comment by ViTsvetkova** [comment]
10. A markup can be applied on a Locality for several agencies 
[in the markup assignment tab User can configure the locality filter (by predictions) and several agencies (by presictions)]

**Comment by mingazhev** [comment]
Suggestions:
1. Leave global markups but change in the following way:
- Markup can be only single (not many as now)
- Can contain only percents 
- Cannot contain any agency location or region filters
- Has a separate set of endpoints and models to manage

2. Add additional tabs and endpoint sets for regional markups. They will support adding markups of 3 types:
- Agency region markup
- Accommodation region markup
- Agency and accommodation region markup

3. Change markup formulas to store actual markup for the agency in a string form (may be divided to multiplication and addition parts)
4. Markups are validated on each change - markup change cannot cause the price to be lower than supplier's
This should be additionally checked (for a case) during markup application while searching
